vignettes_src/cuda.Rmd

```diff
@@ -0,0 +1,326 @@
+  error = FALSE,
+  collapse = TRUE,
+  comment = "#>"
+)
+lang_output <- function(x, lang) {
+  cat(c(sprintf("```%s", lang), x, "```"), sep = "\n")
+}
+cc_output <- function(x) lang_output(x, "cc")
+r_output <- function(x) lang_output(x, "r")
+plain_output <- function(x) lang_output(x, "plain")
+```
+
+With the core approach in dust, you can run models in parallel efficiently up to the number of cores your workstation has available.  Getting more than 32 or 64 cores is hard though, and `dust` provides no multi-node parallelism (e.g., [MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface)). Instead, we have developed a system for running `dust` models on GPUs (graphical processing units), specifically NVIDIA GPUs via CUDA (Compute Unified Device Architecture).
+
+This vignette is written in reverse-order, starting with how to run a model on a GPU, before covering how to write a `dust` model that can be run on a GPU. The reason for this is that we do not expect that people will write these directly; instead we expect that most people will use the [`odin.dust`](https://mrc-ide.github.io/odin.dust/) package to generate these interfaces automatically, without having to write as line of C++.
+
+## Principles
+
+* A GPU model can be run on the CPU
+*
+
+## Running a model with GPU support
+
+The `sirs` model includes GPU support, and will be the focus of this vignette.  However, the installed version cannot be run directly on the GPU for a couple of reasons:
+
+* this would complicate distribution of binaries as we'd depend on all systems having a copy of the CUDA runtime
+* we would have to compile support for many possible GPU architectures at once
+* it is very likely that we'd want to run the model in single precision (`float`) mode rather than double precision (`double`), and changing that requires re-compilation
+
+So here, rather than using `dust::dust_example`, we compile the model and pass in arguments:
+
+```{r}
+path <- system.file("examples/sirs.cpp", package = "dust")
+sirs <- dust::dust(path, gpu = TRUE, real_t = "float")
+```
+
+Notice in compilation that `nvcc` is used to compile the model, rather than `g++`.  The additional option `-gencode=arch=compute_XX,code=sm_XX` was added by `dust` and will include the CUDA compute version supported by your graphics cards. You can use `dust::dust_cuda_options` to set additional options, passing in the return value for the `gpu` argument above.
+
+Once compiled with GPU support, the static method `has_cuda` will report `TRUE`:
+
+```{r}
+sirs$public_methods$has_cuda()
+```
+
+and the static method `device_info` will report on the devices available:
+
+```{r}
+sirs$public_methods$device_info()
+```
+
+If you have more than one GPU, the `id` in the `devices` section will be useful for targeting the correct device.
+
+The object is initialised as usual but with slight differences:
+
+* you will probably want a (much) larger number of particles to take advantage of your device
```

Comment:
```suggestion
* you will probably want a (much) larger number of particles to take advantage of your device. As a rule of thumb we would suggest at least 10000, but depending on model and card 10-100x this may improve performance (relative to the same on a multicore CPU). See below for more discussion of this.
```